### PR TITLE
Move array_poolt class to its own files

### PR DESCRIPTION
**Pull request by romainbrenguier**
This makes it easier to find array_poolt related code.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by smowton — APPROVED** [review]

**Comment by tautschnig** [comment]
@romainbrenguier clang-format, cpplint, doxygen are unhappy.

File: src/solvers/strings/array_pool.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+Module: String solver
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#include "array_pool.h"
+
+/// generate a new symbol expression of the given type with some prefix
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.cpp]: Capitalise first word :bird:

File: src/solvers/strings/array_pool.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+/// \par parameters: a prefix and a type
+/// \return a symbol of type tp whose name starts with "string_refinement#"
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.cpp]: What is "tp"? Should it be replaced by "\p type"

File: src/solvers/strings/array_pool.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  if(s.length() == infinity_exprt(s.length().type()))
+  {
+    auto it = length_of_array.find(s);
+    if(it != length_of_array.end())
+      return it->second;
+  }
+  return s.length();
+}
+
+/// construct a string expression whose length and content are new variables
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.cpp]: :bird:

File: src/solvers/strings/array_pool.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  return pool.find(pointer, length);
+}
+
+array_string_exprt of_argument(array_poolt &array_pool, const exprt &arg)
+{
+  const auto string_argument = expr_checked_cast<struct_exprt>(arg);
+  return array_pool.find(string_argument.op1(), string_argument.op0());
+}
+
+/// casts an expression to a string expression, or fetches the actual
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.cpp]: :bird:

File: src/solvers/strings/array_pool.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+const array_string_exprt &
+array_poolt::find(const exprt &pointer, const exprt &length)
+{
+  const array_typet array_type(pointer.type().subtype(), length);
+  const array_string_exprt array =
+    make_char_array_for_char_pointer(pointer, array_type);
+  return *created_strings_value.insert(array).first;
+}
+
+/// Adds creates a new array if it does not already exists
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.cpp]: Does it add it or create it?

File: src/solvers/strings/array_pool.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+  const array_string_exprt &find(const exprt &pointer, const exprt &length);
+
+  const std::set<array_string_exprt> &created_strings() const;
+
+  array_string_exprt
+  fresh_string(const typet &index_type, const typet &char_type);
+
+private:
+  // associate arrays to char pointers
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.h]: :bird:

Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.h]: `//` -> `///` to trigger doxygen :cat:

File: src/solvers/strings/array_pool.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  std::unordered_map<exprt, array_string_exprt, irep_hash> arrays_of_pointers;
+
+  // associate length to arrays of infinite size
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.h]: :bird: :cat:

File: src/solvers/strings/array_pool.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  std::unordered_map<array_string_exprt, symbol_exprt, irep_hash>
+    length_of_array;
+
+  // generates fresh symbols
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.h]: :bird: :cat:

File: src/solvers/strings/array_pool.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  symbol_generatort &fresh_symbol;
+
+  // Strings created in the pool
```
Comment (owen-mc-diffblue) [inline src/solvers/strings/array_pool.h]: :bird: :cat:

**Review by owen-mc-diffblue — APPROVED** [review]
I've made some suggestions for cleaning up the comments a bit. I appreciate that you're just moving code, so feel free to ignore them if you don't have time.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 876c25d4cbadf1cc852322c6a21978268568b2cd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108499851
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 91349a12a110357b5e0bbf16398a493f22253a96).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108500135

**Review by peterschrammel — APPROVED** [review]

**Comment by tautschnig** [comment]
@romainbrenguier Could you please rebase? Travis has finished, but for some reason GitHub did not get a notification thereof.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4563c75626051c0706d8ef940de2f8639bcd1a27).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108656011

**Pull request merged**